### PR TITLE
Metric Accumulator fix for SumObservers

### LINKED ISSUE #1368: SumObserver sums up observations
otel version: v0.14.0

**Expected behavior**

The OpenTelemetry specification claims:

> Asynchronous, adding instruments (e.g., SumObserver) are used to capture sums directly. This means, for example, that in any sequence of SumObserver observations for a given instrument and label set, the Last Value defines the sum of the instrument.

The comment on Int64SumObserver claims:

```go
// Int64SumObserver is a metric that captures a precomputed sum of
// int64 values at a point in time.
```

I expect that calling `Int64SumObserver.Observe(v)` only returns the latest observation instead of a sum of all observations.

**Actual Behavior**

`Int64SumObserver` sums up all observations instead of using the last value.

This makes it impossible to define Prometheus `COUNTER`s from external sources.
I found this bug when trying to export a counter in Redis.

**Code Example**

```go
package example

import (
	"context"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/metric/prometheus"
	"go.opentelemetry.io/otel/metric"
)

func TestSums(t *testing.T) {
	// Install new Prometheus exporter
	exporter, err := prometheus.InstallNewPipeline(prometheus.Config{})
	require.NoError(t, err)
	// Define a sum observer that always sees "100".
	_, err = otel.Meter("test").NewInt64SumObserver("test_sum", func(ctx context.Context, result metric.Int64ObserverResult) {
		result.Observe(100)
	})
	require.NoError(t, err)
	// Send some requests
	req := &http.Request{
		Method: "GET",
		URL: &url.URL{
			Scheme: "http",
			Host:   "localhost",
			Path:   "/metrics",
		},
	}
	exporter.ServeHTTP(httptest.NewRecorder(), req)
	time.Sleep(15*time.Second)
	exporter.ServeHTTP(httptest.NewRecorder(), req)
	time.Sleep(15*time.Second)
	exporter.ServeHTTP(httptest.NewRecorder(), req)
	time.Sleep(15*time.Second)
	// Print last request
	recorder := httptest.NewRecorder()
	exporter.ServeHTTP(recorder, req)
	body := recorder.Body.String()
	t.Log("Exported Metrics\n" + body)
}

``` 

Outputs

```prometheus
# HELP test_sum 
# TYPE test_sum counter
test_sum 400
```

### PR DESCRIPTION
**Pull request by jmacd**
This bug makes the data from SumObserver and UpDownSumObserver instruments incorrect after the first collection. There were tests for the Processor behavior, which is relatively tricky, but not for the Accumulator. This adds a number of tests.

The means of fixing this is to let `SynchronizedMove(nil)` mean "reset this accumulator" without allocations, which leads to simplifying the Accumulator.

Fixes #1368

### CODE REVIEW

**Review by krnowak — COMMENTED** [review]
One question, otherwise looks good.

File: sdk/metric/aggregator/array/array.go [hunk 1]
```diff
@@ -97,20 +97,27 @@ func (c *Aggregator) Points() ([]number.Number, error) {
...
+	if o != nil {
+		o.points = c.points
+		o.sum = c.sum
+	}
+	c.points = nil
+	c.sum = 0
 	c.lock.Unlock()
 
+	if oa != nil && o == nil {
+		return aggregator.NewInconsistentAggregatorError(c, oa)
```
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/metric/aggregator/array/array.go]: Yes! Thanks for pointing that out. I fixed this in four places and added a new test.
FYI I'm interested in removing one or two aggregators, there are too many :-)

**Review by krnowak — APPROVED** [review]

File: sdk/metric/aggregator/aggregatortest/test.go [hunk 2]
```diff
@@ -16,16 +16,21 @@ package aggregatortest // import "go.opentelemetry.io/otel/sdk/metric/aggregator
 
 import (
 	"context"
+	"errors"
+	"fmt"
```
Comment (krnowak) [inline sdk/metric/aggregator/aggregatortest/test.go:20]: This is unused, which is why CI failed.

**Comment by jmacd** [comment]
@open-telemetry/go-approvers This is a HUGE bug, please approve. Fixes #1368.

**Review by dashpole — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]

File: sdk/metric/sdk.go [hunk 3]
```diff
@@ -182,13 +182,9 @@ func (a *asyncInstrument) observe(num number.Number, labels *label.Set) {
...
 	if ok {
-		if lrec.observedEpoch == a.meter.currentEpoch {
-			// last value wins for Observers, so if we see the same labels
-			// in the current epoch, we replace the old recorder
-			a.meter.processor.AggregatorFor(&a.descriptor, &lrec.observed)
-		} else {
-			lrec.observedEpoch = a.meter.currentEpoch
-		}
+		// Note: SynchronizedMove(nil) can't return an error
+		_ = lrec.observed.SynchronizedMove(nil, &a.descriptor)
```
Comment (jmacd) [inline sdk/metric/sdk.go:186]: Note: This is the line where this bug was fixed.  The `if` branch here was doing a reset, but the `else` branch was not. The code was originally correct and then a bunch of things changed around it, introducing the bug (@krnowak take note!). Specifically, I moved from the Accumulator keeping two Aggregators to keeping one Aggregator, as part of introducing a more-complex Processor component.

The new Processor code was heavily tested, but the former Accumulator wasn't tested enough for the Observer instruments, particularly with multiple rounds. So I had introduced a bug ages ago in the move to a single Aggregator per Accumulator record. We could've fixed this code by calling 
```
a.meter.processor.AggregatorFor(&a.descriptor, &lrec.observed)
```
unconditionally (the `if`-branch), but that's a wasted allocation. Introducing support for resetting the Aggregator fixes the bug w/o the wasted allocation.

Comment (jmacd) [reply sdk/metric/sdk.go:186]: I'm referring to this PR where I think the bug starts. https://github.com/open-telemetry/opentelemetry-go/pull/812

**Comment by jmacd** [comment]
Just for emphasis, I believe this bug has been in place since #812. :-(

**Review by MrAlias — APPROVED** [review]

**Pull request merged**